torch/optim/adam.py

```diff
@@ -244,7 +250,7 @@ def step(self, closure=None):
             adam(
                 params_with_grad,
                 grads,
-                exp_avgs,
+                exp_avgs if beta1 > 0 else grads,
```

Comment:
This is a clever approach, but unfortunately the change will have to be more cautious + spotchecked than this, because we _do_ inplace write into the exp_avgs but we never want to inplace modify the grads.

Reply:
Thank you for your response. For in-place writes, I believe you're referring to `lerp`. Doesn't it keep exp_avgs constant with a coefficient of 1?

In any case, do you have any suggestions on how to proceed? Whether it's to fix this approach or to explore a new one?

Reply:
For proceeding, instead of passing grad into exp_avg indiscriminately, I think we should go to every usage of exp_avg and locally assess what to do if beta1 is 0 to ensure that we don't write into it by accident AND to make sure we're not doing unnecessary work. For example, I would guard the following with

if device_beta1 > 0:
    exp_avg.lerp_(grad, 1 - device_beta1)
else:
    exp_avg = grad

https://github.com/pytorch/pytorch/blob/main/torch/optim/adamw.py#L423

Maybe a good first step before this PR is to write a test making sure grad is never inplaced into...let me do that right now actually!

Reply:
https://github.com/pytorch/pytorch/pull/143612 done here, i'll try merging this soon